.github/workflows/main.yml

```diff
@@ -0,0 +1,44 @@
+# Adapted form https://github.com/avsm/hello-world-action-ocaml/blob/master/.github/workflows/workflow.yml
+name: Main workflow
+
+on:
+  - pull_request
+
+jobs:
+  build:
+    strategy:
+      fail-fast: false
+      matrix:
+        os:
+          - ubuntu-latest
+          # - macos-latest
+          # - windows-latest
+        ocaml-version:
+          - 4.11.0
+          # - 4.10.1
+          # - 4.09.1
+          # - 4.08.1
+
+    runs-on: ${{ matrix.os }}
+
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v2
+
+      - name: Cache opam repository
+        uses: actions/cache@v2
+        with:
+          path: ~/.opam
+          key: ${{ runner.os }}-${{ hashFiles('*.opam') }}-${{ matrix.ocaml-version }}
```

Comment:
Caching the entire switch is not a good idea. By default, setup-ocaml caches everything it can safely cache. (For example: `~/.opam/download-cache`)

Reply:
I see. Thanks for the warning! Any pointers towards something that explains the risks? It seems like it can significantly speed up the cycle, so I'd be interested to know what the tradeoffs are.

Reply:
Currently this is the action/cache itself problem, but it breaks the cache in some cases. Also, caching the entire switch causes the problem of not pulling the latest version of packages unless you explicitly specify the version or use opam lock, which can sometimes delay the discovery of a problem you want to find.

Reply:
The dune-cache makes the opam install process a few times faster for many projects, so it's a better option to use dune-cache instead of caching the whole. (Perhaps triggering the build again will make the process a little faster in this PR from the next job.)

Reply:
I see! All clear now. Thanks so much for the tips here!

Reply:
Well, we can check the consistency of the installed opam packages by calling `opam upgrade --fixup`, but I don't think it's ideal for end-users to do that in all their workflows. So I don't really want to recommend that.

Reply:
Anytime!

Reply:
Yes, this makes sense. This is another thing I'll use at my own risk ;)